File: backend/app/debug_trace.py
import json
import os
import time

DEBUG_LOG_PATH = os.getenv(
    "DEBUG_LOG_PATH",
    "/Users/mahdiabbasian/Documents/Android/Workspace/raybod/.cursor/debug-e7b765.log",
)
SESSION_ID = "e7b765"


def agent_debug_log(
    hypothesis_id: str,
    location: str,
    message: str,
    data: dict | None = None,
    run_id: str = "pre-fix",
) -> None:
    entry = {
        "sessionId": SESSION_ID,
        "hypothesisId": hypothesis_id,
        "location": location,
        "message": message,
        "data": data or {},
        "timestamp": int(time.time() * 1000),
        "runId": run_id,
    }
    line = json.dumps(entry)
    try:
        with open(DEBUG_LOG_PATH, "a", encoding="utf-8") as handle:
            handle.write(line + "\n")
    except OSError:
        pass

File: android/data/src/main/java/dev/abbasian/data/debug/DebugTrace.kt
package dev.abbasian.data.debug

import android.util.Log
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

object DebugTrace {
    private const val TAG = "RaybodDebug"
    private const val SESSION = "e7b765"
    private const val INGEST = "http://127.0.0.1:7249/ingest/fb28033a-9dae-4fbb-9c07-78ad45dfb7cf"
    private val client = OkHttpClient()

    fun log(
        hypothesisId: String,
        location: String,
        message: String,
        data: Map<String, Any?> = emptyMap(),
        runId: String = "post-fix"
    ) {
        val payload = JSONObject().apply {
            put("sessionId", SESSION)
            put("hypothesisId", hypothesisId)
            put("location", location)
            put("message", message)
            put("timestamp", System.currentTimeMillis())
            put("runId", runId)
            put("data", JSONObject(data))
        }
        val line = payload.toString()
        Log.i(TAG, line)
        // #region agent log
        try {
            val body = line.toRequestBody("application/json".toMediaType())
            val request = Request.Builder()
                .url(INGEST)
                .addHeader("X-Debug-Session-Id", SESSION)
                .post(body)
                .build()
            client.newCall(request).enqueue(object : okhttp3.Callback {
                override fun onFailure(call: okhttp3.Call, e: java.io.IOException) {}
                override fun onResponse(call: okhttp3.Call, response: okhttp3.Response) {
                    response.close()
                }
            })
        } catch (_: Exception) {
        }
        // #endregion
    }
}
